### PR TITLE
Add strict liveslots test env with upgrade tools

### LINKED ISSUE #9126: old vow watch not woken up by fulfillment in new incarnation.
Bug isolated and demonstrated at https://github.com/Agoric/agoric-sdk/pull/9125 . 

Isolates the bug that #9097 is stuck on. As isolated, the problem does reproduce.

I'm following the general testing pattern used in #9097, especially in test-async-flow.js, which is where I'm having this problem. There, see https://github.com/Agoric/agoric-sdk/blob/9a00502ec871ee623b88b1781b74d80abd25fdc5/packages/zone/test/async-flow/test-async-flow.js#L181-L182
I shouldn't need the `asyncFlow.wake()` call. But if I remove it I get the same symptom isolated here:

If I `watch` with an unresolved durable vow and durable watcher in the first incarnation, and then I resolve that durable vow in the second incarnation, the durable watcher that was watching that vow should now wake up. It doesn't.

@michaelfig , is this a bug in watching durable vows across incarnations?

### PR DESCRIPTION
**Pull request by mhofman**
closes: #9126

Best reviewed commit-by-commit

## Description
This PR add a new `prepare-strict-test-env.js` which enforces durable requirements on the baggage and provide helpers to simulate upgrades based on the helper from the `async-flow` and `zone` package tests.

In particular it adds a new `startLife` helper which enforces kind redefinition rules for a `build` step, and rejects previously watched promises (assuming they were all decided by the previous incarnation).

### Security Considerations

None, test infra only

### Scaling Considerations

None

### Documentation Considerations

Some types added and clarified. Internal tooling

### Testing Considerations

Adds tests of the new strict env and `startLife` helper.
Add liveslots based test in `vow` using the new `startLife` helper.
Updated the test environments in the `zone` and `async-flow` package to use the new liveslots env, but does not update the `async-flow` tests to use the stricter `startLife` (will be done in #9933 or #9383)

### Upgrade Considerations
Better upgrade testing without needing a full swingset kernel

### CODE REVIEW

File: packages/swingset-liveslots/test/test-env-upgrade.test.js
Comment (michaelfig) [inline packages/swingset-liveslots/test/test-env-upgrade.test.js:1]: We stopped using `test-` as a prefix once we moved to `.test.js`, so it looks a little out-of-place.  Consider renaming this file to `strict-test-env-upgrade.test.js` or something else.

File: packages/swingset-liveslots/test/test-env-upgrade.test.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+import { makeUpgradeDisconnection } from '@agoric/internal/src/upgrade-api.js';
+import { makeExoUtils } from './exo-utils.js';
+
+test.serial('kind redefinition enforced', async t => {
+  annihilate();
+
+  const { prepareExoClass } = makeExoUtils(globalThis.VatData);
+
+  await startLife(async baggage => {
```
Comment (michaelfig) [inline packages/swingset-liveslots/test/test-env-upgrade.test.js]: Oooh... I really love the shape of these `startLife` calls.  As you said, it clearly delineates the prerequisites for the subsequent test code.

File: packages/swingset-liveslots/test/virtual-objects/virtualObjectManager.test.js
Comment (michaelfig) [inline packages/swingset-liveslots/test/virtual-objects/virtualObjectManager.test.js:1]: Nice clarity that the ID creation is not lazy.

File: packages/swingset-liveslots/tools/fakeVirtualSupport.js [hunk 2]
```diff
@@ -339,7 +344,7 @@ export function makeFakeWatchedPromiseManager(
  * @param {object} [options]
  * @param {number} [options.cacheSize]
  * @param {boolean} [options.relaxDurabilityRules]
- * @param {Map<any, any>} [options.fakeStore]
+ * @param {Map<string, string>} [options.fakeStore]
```
Comment (michaelfig) [inline packages/swingset-liveslots/tools/fakeVirtualSupport.js:347]: 💯

File: packages/swingset-liveslots/tools/setup-vat-data.js [hunk 3]
```diff
@@ -44,18 +46,40 @@ globalThis.VatData = harden({
 
 globalThis[PassStyleOfEndowmentSymbol] = passStyleOf;
 
-export const reincarnate = (options = {}) => {
-  const { fakeStore = new Map(), fakeVomKit: fvk } = options;
+/**
+ * @typedef {import("@agoric/internal").Simplify<Omit<NonNullable<Parameters<typeof makeFakeVirtualStuff>[0]>, 'WeakMap' | 'WeakSet'> & {fakeVomKit: FakeVomKit; fakeStore: Map<string, string>}>} ReincarnateOptions
```
Comment (michaelfig) [inline packages/swingset-liveslots/tools/setup-vat-data.js]: Could you please add some newlines in this typedef?  I can't easily see which parts are nested under which other parts.

File: packages/vow/test/watch-upgrade.test.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+      });
+
+      /** @type {import('../src/types.js').VowResolver} */
+      const testVowKitResoler = zone.makeOnce('testVowKit', () => {
+        t.fail('testVowKit maker called');
+      }).resolver;
+
+      return { testVowKitResoler };
+    },
+    async ({ testVowKitResoler }) => {
```
Comment (michaelfig) [inline packages/vow/test/watch-upgrade.test.js]:
```suggestion
      return { testVowKitResolver };
    },
    async ({ testVowKitResolver }) => {
```

File: packages/vow/test/watch-upgrade.test.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        },
+        onRejected(reason) {
+          t.fail(
+            `Second incarnation watcher onRejected triggered with reason ${reason}`,
+          );
+        },
+      });
+
+      /** @type {import('../src/types.js').VowResolver} */
+      const testVowKitResoler = zone.makeOnce('testVowKit', () => {
```
Comment (michaelfig) [inline packages/vow/test/watch-upgrade.test.js]:
```suggestion
      const testVowKitResolver = zone.makeOnce('testVowKit', () => {
```

File: packages/vow/test/watch-upgrade.test.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    const testVowKit = zone.makeOnce('testVowKit', () => vowTools.makeVowKit());
+
+    vowTools.watch(testVowKit.vow, watcher);
+  });
+
+  await startLife(
+    baggage => {
+      const zone = makeDurableZone(baggage, 'durableRoot');
+      prepareVowTools(zone);
+      zone.exo('DurableVowTestWatcher', undefined, {
```
Comment (michaelfig) [inline packages/vow/test/watch-upgrade.test.js:43]: Did you mean to do something with this watcher?  It seems just to be dropped on the floor.

```suggestion
      const watcher = zone.exo('DurableVowTestWatcher', undefined, {
```

**Review by michaelfig — APPROVED** [review]
This is great stuff!  I noticed just some minor details I'd like to see improved.

Approving conditionally assuming those are addressed.

File: packages/vow/test/watch-upgrade.test.js (hunk 6 above)
Comment (mhofman) [reply packages/vow/test/watch-upgrade.test.js:43]: The behavior is redefined. This singleton exo is already setup as watcher in the previous incarnation, so we don't care about its identity here, just that a behavior is reconnected when invoked when resolving the vow.
Added a comment.

**Pull request merged**